### PR TITLE
Drees 102 pmsi flattening

### PR DESCRIPTION
**Pull request by alecoules**
Hi,

Our pull request is ready for the flattening of MCO tables and SSR/RIP tables. The issue was that MCO, SSR and RIP can't be handled in the same way as DCIR. 
For the PMSI, we have to first join the two central tables (MCO_B, MCO_C) and then for each other table join to the result and stack every outputs (code is in **PMSIFlatTable.scala**). 
For SSR and RIP, it's the same thing but we have to be be careful because the keys aren't the same for the first join and the following (code is in **SSRRIPFlatTable.scala**).
We had to add new keys in the conf files : **joinKeysPatient**, **pmsiPatientTableName**. The technique chosen by the algorithm to join the input tables depends of the existence of these keys in the conf file (code is in **FlatteningConfig.scala** and in **Flattening.scala**).
We also wrote tests (**PMSIFlatTableSuite.scala** and **SSRRIPFlatTableSuite.scala**) and add csv, parquet and conf files for these tests.
We tested our code on the Drees sample (4M persons during 9 years).

Arnaud Lecoules

### CODE REVIEW

File: src/main/scala/fr/polytechnique/cmap/cnam/flattening/PMSIFlatTable.scala [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+  extends FlatTable(sqlContext: SQLContext, config: JoinTableConfig) {
+
+  val pmsiPatientTable: Table = Table.build(sqlContext, inputBasePath, config.pmsiPatientTableName.get)
+
+  /**
+   * This method merge two schemas, if a column is not in the schema of the
+   * DataFrame, it created the column as empty
+   */
+  def mergeSchemas(myCols: Set[String], allCols: Set[String]) = {
+    allCols.toList.map(x => x match {
```
Comment (ysebiat) [inline src/main/scala/fr/polytechnique/cmap/cnam/flattening/PMSIFlatTable.scala]: The variable x is ambiguous in this method. The name x captures the element you pass. Here is a suggestion:
```scala
  def mergeSchemas(myCols: Set[String], allCols: Set[String]): List[Column] = 
    allCols.toList.map(colName => if (myCols.contains(colName)) col(colName) else lit(null).alias(colName))
```

**Review by ysebiat — CHANGES_REQUESTED** [review]
A comment on the merge method.

File: src/main/scala/fr/polytechnique/cmap/cnam/flattening/PMSIFlatTable.scala (hunk 1 above)
Comment (alecoules) [reply src/main/scala/fr/polytechnique/cmap/cnam/flattening/PMSIFlatTable.scala]: OK. I commited this change.

Comment (alecoules) [reply src/main/scala/fr/polytechnique/cmap/cnam/flattening/PMSIFlatTable.scala]: Subsidiary question : why did you prefere if/else versus match pattern ?

Comment (ysebiat) [reply src/main/scala/fr/polytechnique/cmap/cnam/flattening/PMSIFlatTable.scala]: I would prefer the match pattern for complex pattern matching. The example above is a boolean if around a test, which IMO is better suited for an if/else control.

File: src/main/resources/flattening/config/main.conf [hunk 2]
```diff
@@ -17,3 +17,39 @@ pio = {
 fraude = {
   include "fraude/env.conf"
 }
+
```
Comment (ysebiat) [inline src/main/resources/flattening/config/main.conf:20]: Do you think this configurations are necessary for the whole project?

File: src/main/resources/flattening/config/ssr_month/env.conf [hunk 3]
```diff
@@ -0,0 +1,23 @@
+base_path = "target/test/output"
```
Comment (ysebiat) [inline src/main/resources/flattening/config/ssr_month/env.conf:1]: I think you can omit this file from the PR.

File: src/main/resources/flattening/config/ssr_month/ssr.conf [hunk 4]
```diff
@@ -0,0 +1,67 @@
+tables = [
```
Comment (ysebiat) [inline src/main/resources/flattening/config/ssr_month/ssr.conf:1]: I think you can omit this file from the PR.

File: src/main/scala/fr/polytechnique/cmap/cnam/flattening/Flattening.scala [hunk 5]
```diff
@@ -46,11 +46,33 @@ object Flattening {
       }.toList
   }
 
+  /**
```
Comment (ysebiat) [inline src/main/scala/fr/polytechnique/cmap/cnam/flattening/Flattening.scala:49]: Add documentation for the arguments.

File: src/main/scala/fr/polytechnique/cmap/cnam/flattening/Flattening.scala [hunk 6]
```diff
@@ -46,11 +46,33 @@ object Flattening {
...
+        OperationMetadata(config.name, config.flatOutputPath.get, "flat_table", config.mainTableName :: config.tablesToJoin, config.joinKeys)
+      }
+      else {
+        if (config.joinKeysPatient.isEmpty) {
+          logger.info("join table " + config.name + " with Pmsi logic")
+          new PMSIFlatTable(sqlContext, config).writeAsParquet()
+          OperationMetadata(config.name, config.flatOutputPath.get, "flat_table", config.mainTableName :: config.tablesToJoin, config.joinKeys)
+        }
+        else {
+          logger.info("join SSR table " + config.name + " with Pmsi logic")
```
Comment (ysebiat) [inline src/main/scala/fr/polytechnique/cmap/cnam/flattening/Flattening.scala]: Better to say
```scala
logger.info("join table " + config.name + " with SSR-PMSI logic")
```

File: src/test/scala/fr/polytechnique/cmap/cnam/flattening/FlatteningSuite.scala [hunk 7]
```diff
@@ -5,6 +5,7 @@ package fr.polytechnique.cmap.cnam.flattening
 import org.apache.spark.sql._
 import fr.polytechnique.cmap.cnam.SharedContext
 import fr.polytechnique.cmap.cnam.utilities.DFUtils._
+import org.apache.spark.sql.functions.col
```
Comment (ysebiat) [inline src/test/scala/fr/polytechnique/cmap/cnam/flattening/FlatteningSuite.scala]: The order of the import must follow the following order:
```scala
java
javax
scala
all other imports
org.apache.spark
fr.polytechnique

File: src/test/scala/fr/polytechnique/cmap/cnam/flattening/PMSIFlatTableSuite.scala [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+import fr.polytechnique.cmap.cnam.flattening.FlatteningConfig.YearAndMonths
+import fr.polytechnique.cmap.cnam.utilities.DFUtils._
+import fr.polytechnique.cmap.cnam.utilities.Functions._
+import org.apache.spark.sql.DataFrame
+import org.mockito.Mockito._
+import org.apache.spark.sql.functions._
+
+class PMSIFlatTableSuite extends SharedContext {
+
+  "WriteAsParquet" should "flatten MCO and write it in the correct path" in {
```
Comment (ysebiat) [inline src/test/scala/fr/polytechnique/cmap/cnam/flattening/PMSIFlatTableSuite.scala]: ```writeAsParquet```

File: src/test/scala/fr/polytechnique/cmap/cnam/flattening/PMSIFlatTableSuite.scala [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+    result_cols = result_cols//.filter(column => !(column.contains("GHM")))
+    val result_sorted = result.select(result_cols.map(col):_*)
+
+    // Then
+    assert(resultPath == flattenedTableTest.outputBasePath)
+    assert(expectedDF_sorted.columns.toSet == result_sorted.columns.toSet)
+    assert(expectedDF_sorted.count() == result_sorted.count())
+  }
+
+  "WriteAsParquetMonth" should "flatten MCO monthly and write it in the correct path" in {
```
Comment (ysebiat) [inline src/test/scala/fr/polytechnique/cmap/cnam/flattening/PMSIFlatTableSuite.scala]: ```writeAsParquetMonth```

**Review by ysebiat — CHANGES_REQUESTED** [review]
Some remarks about
1. Order of the imports.
2. Some configuration files that I think are not necessary for the PR.

File: src/main/resources/flattening/config/main.conf (hunk 2 above)
Comment (alecoules) [reply src/main/resources/flattening/config/main.conf:20]: I dropped every confs related to drees_echantillon

File: src/main/resources/flattening/config/ssr_month/env.conf (hunk 3 above)
Comment (alecoules) [reply src/main/resources/flattening/config/ssr_month/env.conf:1]: I'm using it for the tests

File: src/main/resources/flattening/config/ssr_month/ssr.conf (hunk 4 above)
Comment (alecoules) [reply src/main/resources/flattening/config/ssr_month/ssr.conf:1]: I'm using it for the tests

File: src/main/scala/fr/polytechnique/cmap/cnam/flattening/Flattening.scala (hunk 5 above)
Comment (alecoules) [reply src/main/scala/fr/polytechnique/cmap/cnam/flattening/Flattening.scala:49]: Done

**Review by ysebiat — APPROVED** [review]
Ready to merge.

**Pull request merged**